diff --git a/Makefile b/Makefile
index 93de6db..e63452a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #configureable stuff
-PREFIX ?= /usr/local
+PREFIX ?= /usr
 BINDIR ?= $(PREFIX)/bin
 MANDIR ?= $(PREFIX)/share/man/man1
 LIBDIR ?= $(PREFIX)/lib/sendip
@@ -8,7 +8,7 @@ INSTALL ?= install
 #For Solaris, you may need
 #INSTALL=/usr/ucb/install
 
-CFLAGS=	-fPIC -fsigned-char -pipe -Wall -Wpointer-arith -Wwrite-strings \
+CFLAGS +=	-fPIC -fsigned-char -pipe -Wall -Wpointer-arith -Wwrite-strings \
 			-Wstrict-prototypes -Wnested-externs -Winline -Werror \
 			-Wcast-align -O3 \
 			-DSENDIP_LIBS=\"$(LIBDIR)\"
@@ -32,7 +32,7 @@ LIBS= libsendipaux.a
 LIBOBJS= csum.o compact.o protoname.o headers.o parseargs.o cryptomod.o crc32.o filearray.o
 SUBDIRS= mec
 
-all:	$(LIBS) subdirs sendip $(PROTOS) sendip.1 sendip.spec sendipman.html
+all:	$(LIBS) subdirs sendip $(PROTOS)
 
 #there has to be a nice way to do this
 sendip:	sendip.o	gnugetopt.o gnugetopt1.o compact.o filearray.o
diff --git a/mec/Makefile b/mec/Makefile
index 128c5ee..42d07d5 100644
--- a/mec/Makefile
+++ b/mec/Makefile
@@ -1,5 +1,5 @@
 #configureable stuff
-PREFIX ?= /usr/local
+PREFIX ?= /usr
 BINDIR ?= $(PREFIX)/bin
 MANDIR ?= $(PREFIX)/share/man/man1
 LIBDIR ?= $(PREFIX)/lib/sendip
@@ -8,7 +8,7 @@ INSTALL ?= install
 #For Solaris, you may need
 #INSTALL=/usr/ucb/install
 
-CFLAGS=	-I.. -fPIC -fsigned-char -pipe -Wall -Wpointer-arith -Wwrite-strings \
+CFLAGS +=	-I.. -fPIC -fsigned-char -pipe -Wall -Wpointer-arith -Wwrite-strings \
 	-Wstrict-prototypes -Wnested-externs -Winline -Werror -O3 -Wcast-align \
 	-DSENDIP_LIBS=\"$(LIBDIR)\"
 #-Wcast-align causes problems on solaris, but not serious ones
