diff --git a/target/linux/x86/config-4.14 b/target/linux/x86/config-4.14
index e2734e3f7e..c553316ff5 100644
--- a/target/linux/x86/config-4.14
+++ b/target/linux/x86/config-4.14
@@ -391,7 +391,7 @@ CONFIG_RATIONAL=y
 # CONFIG_RCU_STALL_COMMON is not set
 CONFIG_RD_BZIP2=y
 CONFIG_RD_GZIP=y
-CONFIG_RETPOLINE=y
+# CONFIG_RETPOLINE is not set
 CONFIG_RTC_CLASS=y
 CONFIG_RTC_MC146818_LIB=y
 CONFIG_RWSEM_XCHGADD_ALGORITHM=y
