diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1836e61..22d0aa6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,11 +10,11 @@ project(KinesisVideoProducerCpp)
 set(CMAKE_CXX_STANDARD 11)
 
 # User Flags
-option(BUILD_GSTREAMER_PLUGIN "Build kvssink GStreamer plugin" OFF)
+option(BUILD_GSTREAMER_PLUGIN "Build kvssink GStreamer plugin" ON)
 option(BUILD_JNI "Build C++ wrapper for JNI to expose the functionality to Java/Android" OFF)
 option(BUILD_STATIC "Build with static linkage" OFF)
 option(ADD_MUCLIBC "Add -muclibc c flag" OFF)
-option(BUILD_DEPENDENCIES "Whether or not to build depending libraries from source" ON)
+option(BUILD_DEPENDENCIES "Whether or not to build depending libraries from source" OFF)
 option(BUILD_OPENSSL_PLATFORM "If buildng OpenSSL what is the target platform" OFF)
 
 # Developer Flags
